api/src/stampy_chat/chat.py

```diff
@@ -186,22 +139,24 @@ def talk_to_robot_internal(index, query: str, mode: str, history: Prompt, sessio
 
         # 2. Generate a prompt
         yield {"state": "loading", "phase": "prompt"}
-        prompt = construct_prompt(query, mode, history, top_k_blocks)
+        prompt = construct_prompt(query, settings, history, top_k_blocks)
 
         # 3. Run both the standalone query and the full prompt through
         # moderation to see if it will be accepted by OpenAI's api
         check_openai_moderation(prompt, query)
 
         # 4. Count number of tokens left for completion (-50 for a buffer)
-        max_tokens_completion = remaining_tokens(prompt)
+        max_tokens_completion = remaining_tokens(prompt, settings)
+        if max_tokens_completion < 40:
+            raise ValueError(f"{max_tokens_completion} tokens left for the actual query after constructing the context - aborting, as that's not going to be enough")
```

Comment:
should this be changed? I arbitrarily put 40, but 100 would probably also be too few

Reply:
maybe it should be a class or instance variable? Also, What is the trade off here? More tokens cost more per response potentially, but has a better chance of giving the user what we want?

Reply:
There's already a `settings.numTokens` setting which should handle that. The problem here is that a whole lot of additional context is added to the prompt before it gets to this point. Specifially:
1. `settings.source_prompt_prefix`
2. up to `settings.topKBlocks` of citations from pinecone
3. `settings.source_prompt_suffix`
4. (optional) the last `settings.maxHistory` items from the conversation
5. `settings.question_prompt`
6.  (optional) `settings.mode_prompt` to regulate the complexity level of the answer
7. the actual query

This prompt can get quite big, taking up a lot of the `settings.numTokens` available tokens (in the worst case taking all of them), which limits the number left for the actual response. A seperate setting might help, but won't solve the underlying problem, which would probably require some creative book keeping to limit the number of tokens available for each step. Steps 1, 3, 5, 6 and 7 should always be sent to the LLM (but even then can take up all the tokens), so it's really only the context and history steps that should be limited.